apps/openmw/mwphysics/movementsolver.cpp

```diff
@@ -167,6 +193,45 @@ namespace MWPhysics
             static const float fStromWalkMult = MWBase::Environment::get().getWorld()->getStore().get<ESM::GameSetting>().find("fStromWalkMult")->mValue.getFloat();
             velocity *= 1.f-(fStromWalkMult * (angleDegrees/180.f));
         }
+        
+        auto tempPosition = physicActor->getPosition();
+        auto refPosition = position - osg::Vec3f(0.0, 0.0, halfExtents.z());
+        physicActor->setPosition(refPosition);
+        // try to pop outside of the world before doing anything else if we're inside of it
+        DeepestContactResultCallback contactCallback{physicActor->getCollisionObject(), velocity};
+        const_cast<btCollisionWorld*>(collisionWorld)->contactTest(physicActor->getCollisionObject(), contactCallback);
+        if(contactCallback.mDistance < -sAllowedPenetration)
+        {
+            bool giveup = false;
+            auto positionDelta = contactCallback.mDelta;
+            physicActor->setPosition(refPosition + Misc::Convert::toOsg(positionDelta));
+
+            DeepestContactResultCallback contactCallback2{physicActor->getCollisionObject(), velocity};
+            const_cast<btCollisionWorld*>(collisionWorld)->contactTest(physicActor->getCollisionObject(), contactCallback2);
+            // try again but only upwards (fixes some bad coc floors)
+            if(contactCallback2.mDistance < contactCallback.mDistance)
+            {
+                physicActor->setPosition(refPosition + osg::Vec3f(0.0, 0.0, fabsf(positionDelta.z())));
+
+                DeepestContactResultCallback contactCallback3{physicActor->getCollisionObject(), velocity};
+                const_cast<btCollisionWorld*>(collisionWorld)->contactTest(physicActor->getCollisionObject(), contactCallback3);
+                // try again but fixed distance
+                if(contactCallback3.mDistance < contactCallback.mDistance)
+                {
+                    physicActor->setPosition(refPosition + osg::Vec3f(0.0, 0.0, 10));
+
+                    DeepestContactResultCallback contactCallback4{physicActor->getCollisionObject(), velocity};
+                    const_cast<btCollisionWorld*>(collisionWorld)->contactTest(physicActor->getCollisionObject(), contactCallback4);
+                    // give up
+                    if(contactCallback4.mDistance < contactCallback.mDistance)
+                        giveup = true;
+                }
+            }
+            if(!giveup)
+                position = physicActor->getPosition() + osg::Vec3f(0.0, 0.0, halfExtents.z());
+            
+        }
+        physicActor->setPosition(tempPosition);
```

Comment:
This whole chunk is problematic to run in background thread. One can't call `setWorldTransform` (called by `setPosition`) from different threads.

I think the same can be achieved with a serie of `convexSweepTest`, ie instead of 

```
contactTest()
if (bad)
{
    setPosition(orig + delta)
    contactTest()
    if (bad again)
    {
        setPosition(orig + delta2)
        contactTest()
    }
}
```
 something like
```
contactTest()
if (bad)
{
    convexSweepTest(actorShape, orig, orig + delta)
    if (bad again)
    {
        convexSweepTest(actorShape, orig, orig + delta2)
    }
}
```

I don't know how practical or if it creates other issues, but that is safe to use at least.

Reply:
Is it just the setWorldTransform that is the issue?

Reply:
I can't use convexSweepTest for this because it misbehaves when you're already inside of geometry, which is what this code is concerned with.

Reply:
> Is it just the setWorldTransform that is the issue?

Here yes. 

 
> I can't use convexSweepTest for this because it misbehaves when you're already inside of geometry, which is what this code is concerned with.

What about putting this in a separate `popOutActor` function and change the algorithm from:

```
doSomePreparation()
if (isStuckInsideSomething)
{
    popOutActor()
}
doTheActualMove()
```
to
```
doSomePreparation()
if (isStuckInsideSomething)
{
    flagActorAsStuck()
    return
}
doTheActualMove()
```

Then I could check the `isStuckInsideSomething` flag after each call to `move()` and call `popOutActor` in a single thread context. This is how I call `setPosition` now, so it could be done in the same place.
The only downside I can think of, is that one simulation step will be skipped for a stuck actor, but given that is an exceptional / abnormal state it may not be that bad. What do you think?

Reply:
That still won't work. The actor would have to be able to recover its exact original position if any of the tests don't work as desired. It probably makes the most sense to do all of this as a separate pass over all actors outside of the movement solver.

Reply:
> That still won't work. The actor would have to be able to recover its exact original position if any of the tests don't work as desired. 

I don't see why it is an issue, it is enough to save and restore the position in case of failure, just as you do now. That being said:

> It probably makes the most sense to do all of this as a separate pass over all actors outside of the movement solver.

Yep, that's a better idea. It could be simply moved to the beginning of `applyQueuedMovement` (or its equivalent). And if it happens to be too expensive (which I doubt) and needs to be moved from the main thread to the background thread it can be done later.